File: customview/src/main/java/com/example/customview/adapter/HoverItemAdapter2.java
package com.example.customview.adapter;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.customview.R;

import java.util.List;

@SuppressLint("RecyclerView")
public class HoverItemAdapter2 extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final String TAG = "MainAdapter";
    private final int type_one = 1;
    private final int type_two = 2;
    private final int type_three = 3;
    private final List<String> mDatas;

    public interface OnItemClickListener {
        void onItemClick(View view, int position);
    }

    public OnItemClickListener mOnItemClickListener;

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.mOnItemClickListener = listener;
    }

    public HoverItemAdapter2(List<String> mDatas) {
        this.mDatas = mDatas;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;
        if (viewType == type_one) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.hover_item_one, parent, false);
            return new OneViewHolder(view);
        } else if (viewType == type_two) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.hover_item_two, parent, false);
            return new TwoViewHolder(view);
        } else {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.hover_item_three, parent, false);
            return new ThreeViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {
        switch (holder.getItemViewType()) {
            case type_one:
                OneViewHolder oneViewHolder = (OneViewHolder) holder;
                oneViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Log.i(TAG, "onItemClick: 点击了 " + position);
                    }
                });
                break;
            case type_two:
                TwoViewHolder twoViewHolder = (TwoViewHolder) holder;
                twoViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        mOnItemClickListener.onItemClick(view, position);
                    }
                });
                break;
            case type_three:
                ThreeViewHolder threeViewHolder = (ThreeViewHolder) holder;
                threeViewHolder.tv_text.setText(mDatas.get(position));
                threeViewHolder.tv_text.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Log.i(TAG, "onItemClick: 点击了 " + position);
                    }
                });
                break;
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position < 3) {
            return type_one;
        } else if (position == 3) {
            return type_two;
        } else {
            return type_three;
        }
    }

    @Override
    public int getItemCount() {
        return mDatas.size();
    }

    static class OneViewHolder extends RecyclerView.ViewHolder {
        public OneViewHolder(View itemView) {
            super(itemView);
        }
    }

    static class TwoViewHolder extends RecyclerView.ViewHolder {
        public TwoViewHolder(View itemView) {
            super(itemView);
        }
    }

    static class ThreeViewHolder extends RecyclerView.ViewHolder {
        TextView tv_text;

        public ThreeViewHolder(View itemView) {
            super(itemView);
            tv_text = itemView.findViewById(R.id.tv_text);
        }
    }
}

File: customview/src/main/java/com/example/customview/adapter/HoverItemAdapter2.kt
package com.example.customview.adapter

import android.annotation.SuppressLint
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.customview.R
import com.example.customview.databinding.HoverItemThreeBinding

/**
 * 封装 Hover Item Adapter 2 相关逻辑的类。
 */
@SuppressLint("RecyclerView")
open class HoverItemAdapter2(private val mDatas: List<String>?) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val TAG = "MainAdapter"
        private const val TYPE_ONE = 1
        private const val TYPE_TWO = 2
        private const val TYPE_THREE = 3
    }

    interface OnItemClickListener {
        fun onItemClick(view: View, position: Int)
    }
    
    private var mOnItemClickListener: OnItemClickListener? = null

    fun setOnItemClickListener(listener: OnItemClickListener) {
        mOnItemClickListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view: View
        return when (viewType) {
            TYPE_ONE -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.hover_item_one, parent, false)
                OneViewHolder(view)
            }

            TYPE_TWO -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.hover_item_two, parent, false)
                TwoViewHolder(view)
            }

            else -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.hover_item_three, parent, false)
                ThreeViewHolder(view)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            TYPE_ONE -> {
                val oneViewHolder = holder as OneViewHolder
                oneViewHolder.itemView.setOnClickListener {
                    Log.i(TAG, "onItemClick: 点击了 $position")
                }
            }

            TYPE_TWO -> {
                val twoViewHolder = holder as TwoViewHolder
                twoViewHolder.itemView.setOnClickListener { view ->
                   mOnItemClickListener?.onItemClick(view, position)
                }
            }

            TYPE_THREE -> {
                val threeViewHolder = holder as ThreeViewHolder
                threeViewHolder.tv_text.text = mDatas!![position]
                threeViewHolder.tv_text.setOnClickListener {
                    Log.i(TAG, "onItemClick: 点击了 $position")
                }
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position < 3) {
            TYPE_ONE
        } else if (position == 3) {
            TYPE_TWO
        } else {
            TYPE_THREE
        }
    }

    override fun getItemCount(): Int {
        return mDatas!!.size
    }

    class OneViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    class TwoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    class ThreeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tv_text: TextView = HoverItemThreeBinding.bind(itemView).tvText
    }
}
